pkg/mev/send_bundle.go

```diff
@@ -14,54 +14,113 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 )
 
-// SendBundle https://docs.flashbots.net/flashbots-auction/advanced/rpc-endpoint#eth_sendbundle,
-// https://beaverbuild.org/docs.html, https://rsync-builder.xyz/docs
-func SendBundle( // nolint: cyclop
-	ctx context.Context, c *http.Client, endpoint string, param *SendBundleParams, options ...SendBundleOption,
-) (SendBundleResponse, error) {
-	var opts sendBundleOpts
-	for _, fn := range options {
-		if fn == nil {
-			continue
-		}
-		fn(&opts)
+// FlashbotBundleSender https://docs.flashbots.net/flashbots-auction/advanced/rpc-endpoint#eth_sendbundle,
+type FlashbotBundleSender struct {
+	c        *http.Client
+	key      *ecdsa.PrivateKey
+	endpoint string
+}
+
+func NewFlashbotSender(
+	c *http.Client,
+	key *ecdsa.PrivateKey,
+	endpoint string,
+) *FlashbotBundleSender {
+	return &FlashbotBundleSender{
+		c:        c,
+		key:      key,
+		endpoint: endpoint,
 	}
+}
 
+func (s *FlashbotBundleSender) Send(
+	ctx context.Context, blockNumber uint64, txs ...*types.Transaction,
+) (SendBundleResponse, error) {
 	req := SendBundleRequest{
 		ID:      SendBundleID,
 		JSONRPC: JSONRPC2,
 		Method:  ETHSendBundleMethod,
 	}
-	req.Params = append(req.Params, param)
+	p := new(SendBundleParams).SetBlockNumber(blockNumber).SetTransactions(txs...)
+	req.Params = append(req.Params, p)
 
 	reqBody, err := json.Marshal(req)
 	if err != nil {
 		return SendBundleResponse{}, fmt.Errorf("marshal json error: %w", err)
 	}
 
-	headers := make(map[string]string)
-	if opts.flashbotSignKey != nil {
-		signature, err := signRequest(opts.flashbotSignKey, reqBody)
-		if err != nil {
-			return SendBundleResponse{}, fmt.Errorf("sign flashbot request error: %w", err)
-		}
+	signature, err := signRequest(s.key, reqBody)
+	if err != nil {
+		return SendBundleResponse{}, fmt.Errorf("sign flashbot request error: %w", err)
+	}
+	flashbotSig := fmt.Sprintf("%s:%s",
+		crypto.PubkeyToAddress(s.key.PublicKey), hexutil.Encode(signature))
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, s.endpoint, bytes.NewBuffer(reqBody))
+	if err != nil {
+		return SendBundleResponse{}, fmt.Errorf("new http request error: %w", err)
+	}
+
+	resp, err := doRequest[SendBundleResponse](
+		s.c,
+		httpReq,
+		[2]string{"X-Flashbots-Signature", flashbotSig})
+	if err != nil {
+		return SendBundleResponse{}, err
+	}
+
+	if len(resp.Error.Messange) != 0 {
+		return SendBundleResponse{}, fmt.Errorf("response error, code: [%d], message: [%s]",
+			resp.Error.Code, resp.Error.Messange)
+	}
+
+	return resp, nil
+}
+
+// BundleSender https://beaverbuild.org/docs.html; https://rsync-builder.xyz/docs
+type BundleSender struct {
```

Comment:
```suggestion
type MEVClient struct {
```

I wonder if we can define `MEVClient` as `FlashbotsClient` with optional key, so if the key is missing we don't need to sign the request.

Reply:
updated in https://github.com/KyberNetwork/tradinglib/pull/17/commits/3ab36734177a6e363d98e855bfe41d3a8bbcf4e1